service/common/src/main/java/org/apache/polaris/service/admin/PolarisAdminService.java

```diff
@@ -200,6 +205,43 @@ private Optional<CatalogRoleEntity> findCatalogRoleByName(String catalogName, St
         .map(path -> CatalogRoleEntity.of(path.getRawLeafEntity()));
   }
 
+  private <T> Stream<T> loadEntities(
+      @Nonnull PolarisEntityType entityType,
+      @Nonnull PolarisEntitySubType entitySubType,
+      @Nullable PolarisEntity catalogEntity,
+      @Nonnull Function<PolarisBaseEntity, T> transformer) {
+    List<PolarisEntityCore> catalogPath;
+    long catalogId;
+    if (catalogEntity == null) {
+      catalogPath = null;
+      catalogId = 0;
+    } else {
+      catalogPath = PolarisEntity.toCoreList(List.of(catalogEntity));
+      catalogId = catalogEntity.getId();
+    }
+    // TODO: add loadEntities method to PolarisMetaStoreManager
+    // loadEntity may return null due to multiple non-atomic API calls to the persistence layer.
+    // Specifically, this can happen when a PolarisEntity is returned by listEntities, but cannot be
+    // loaded afterward because it was purged by another process before it could be loaded.
+    return metaStoreManager
+        .listEntities(
+            getCurrentPolarisContext(),
+            catalogPath,
+            entityType,
+            entitySubType,
+            PageToken.readEverything())
+        .getEntities()
+        .stream()
+        .map(
+            nameAndId ->
+                metaStoreManager.loadEntity(
+                    getCurrentPolarisContext(), catalogId, nameAndId.getId(), nameAndId.getType()))
```

Comment:
Just a quick question, is there a specific reason we need to save the `catalogId` earlier? I'm asking because `nameAndId` already provides `getCatalogId()`.

Reply:
it might be true but i am only following what the existing code was doing... whether `getCatalogId()` always returns the right value for all types of entities idk, so i was just sticking to the existing code

Reply:
Thanks for clarifying it!